doc/design/context.md

```diff
@@ -0,0 +1,175 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run`, which will take `Variable` (containing `Tensor`) from `Scope` to make computation. There will be single or several threads to execute operator's `Run` method.
+
+Since the Tensor computation are executed by Eigen library, which needs an Eigen::GpuDevice type object as  parameter in a GPU card, and the GpuDevice parameter is constructed with an Eigen::CudaStreamDevice object. We need to set both a specific GpuID and CudaStream to create a Eigen::CudaStream object.
+
+At the same time, some computation work will executed by cuBLAS or cuDNN library. Take cuBLAS library as an example, we have to acquire a cublasHandle which binds on a CudaStream to make computation. It's the same way as Eigen library does.
+
+A `Context` is corresponded to a thread and holds runtime resources, such as CudaStream, cublasHandle and so on. And the `Run` method of `Operator` will take `Context` from a specific thread as a parameter to get necessary runtime resources.
```

Comment:
The `Context` here is actually thread Context. The Net is executed by single or several threads, and each thread will have its own runtime resources. Maybe ThreadContext is a better name to understand.